### PR TITLE
[Run2_2017] Transform nested vector branches to flattend+offset vector branches

### PR DESCRIPTION
**Pull request by aperloff**
Allow the transformation of vector<vector<T>> branches to be saved as flattened vector<T> branches with an associated vector<int> branch of offsets.

Two unit tests were performed, one keeping the nested vector format and the other flattening the vectors. Flattening the vectors has the effect of decreasing the ntuple size by a little under 1% (0.88% in our 100 event test).

I have not tested the speed change.

### CODE REVIEW

File: TreeMaker/interface/TreeMaker.h [hunk 1]
```diff
@@ -356,3 +358,83 @@ class TreeRecoCand : public TreeObject<vector<TLorentzVector> > {
 		bool doLorentz{};
 		vector<double> pt, eta, phi, energy;
 };
+
+//derived version of vector<vector<T>>
+//with switch for vector<T> values and vector<int> offsets instead
+template <template<typename...> class R=std::vector, typename Top = R<R<double>>, typename Sub = typename Top::value_type>
```
Comment (kpedro88) [inline TreeMaker/interface/TreeMaker.h]: in what case would R not be std::vector?

Comment (kpedro88) [reply TreeMaker/interface/TreeMaker.h]: wouldn't we prefer to just make the inner type a template, and then build up Sub and Top from that?

File: TreeMaker/interface/TreeMaker.h [hunk 2]
```diff
@@ -356,3 +358,83 @@ class TreeRecoCand : public TreeObject<vector<TLorentzVector> > {
...
+			else{
+				values.clear();
+				offsets.clear();
+			}
+		}
+
+	protected:
+		//member variables
+		edm::EDGetTokenT<Top> tok;
+		string type{};
```
Comment (kpedro88) [inline TreeMaker/interface/TreeMaker.h]: default initialization with `{}` is not necessary for strings (string class is automatically initialized to an empty string)

File: TreeMaker/python/maker.py [hunk 3]
```diff
@@ -146,6 +147,7 @@ def printSetup(self):
         print " Storing a minimal set of GenParticles: "+str(self.saveMinimalGenParticles)
         print " Storing the GenTops: "+str(self.saveGenTops)
         print " Saving the MT2 variable: "+str(self.doMT2)
+        print " Saving nested vectors as vector<vector<T>>: "+str(self.nestedVectors)
```
Comment (kpedro88) [inline TreeMaker/python/maker.py]: i would make this message conditional:
```python
if self.nestedVectors: print " Saving nested vectors as vector<vector<T>>"
else: print " Saving nested vectors as vector<T> + vector<int>"
```

File: TreeMaker/interface/TreeMaker.h (hunk 1 above)
Comment (aperloff) [reply TreeMaker/interface/TreeMaker.h]: 1) I guess in our case we should restrict R to be std::vector. I took this code from somewhere else and if you look at the flatten function it technically doesn't require the outer object to be a std::vector. Do you want me to work on removing R?
2) Doing (1) would make this possible, but is it really necessary? Seems to be roughly equivalent code.

File: TreeMaker/interface/TreeMaker.h (hunk 2 above)
Comment (aperloff) [reply TreeMaker/interface/TreeMaker.h]: Changed

File: TreeMaker/interface/TreeMaker.h (hunk 1 above)
Comment (kpedro88) [reply TreeMaker/interface/TreeMaker.h]: I think it's preferable to make the code as simple as possible (or in other words, only as complex as necessary). It's simpler to specify one parameter, the inner type, and then just typedef Sub and Top as vectors of that inner type. The use of value_type is clever, but unnecessarily so.

File: TreeMaker/python/maker.py (hunk 3 above)
Comment (aperloff) [reply TreeMaker/python/maker.py]: changed

File: TreeMaker/interface/TreeMaker.h (hunk 1 above)
Comment (aperloff) [reply TreeMaker/interface/TreeMaker.h]: Okay, the new commit should make things simpler.

File: TreeMaker/interface/TreeMaker.h [hunk 4]
```diff
@@ -356,3 +358,110 @@ class TreeRecoCand : public TreeObject<vector<TLorentzVector> > {
 		bool doLorentz{};
 		vector<double> pt, eta, phi, energy;
 };
+
+// Derived version of vector<vector<T>> with switch for vector<T> values and vector<int> offsets instead
+template <typename Base = double, typename Sub = std::vector<Base>, typename Top = std::vector<Sub>>
```
Comment (kpedro88) [inline TreeMaker/interface/TreeMaker.h]: in what case would we want Sub *not* to be `std::vector<Base>`?

I think Sub and Top should not be template parameters, but typedefs inside the class:
```c++
typedef Sub std::vector<Base>;
typedef Top std::vector<Sub>;
```

Comment (kpedro88) [reply TreeMaker/interface/TreeMaker.h]: this means you'll have to specify ` : public TreeObject<std::vector<std::vector<Base>>>` in the class definition, but oh well

Comment (aperloff) [reply TreeMaker/interface/TreeMaker.h]: Done

**Pull request merged**